File: 39-combination-sum/39-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        
        result = []
        
        
        def backtrack(i,curr,total):
            
            if total == target:
                result.append(curr[::])
                return 
            
            if i >= len(candidates) or total > target:
                return
            
            total += candidates[i]
            curr.append(candidates[i])
            backtrack(i,curr,total)
            total -= candidates[i]
            curr.pop()
            backtrack(i+1,curr,total)
            
            
        backtrack(0,[],0)
            
        return result
                    

File: 39-combination-sum/39-combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    
    result = []
    
    
    const backtrack = (i=0,curr=[],total=0)=>{
        
        if(total==target){
            result.push([...curr])
            return
        }
        if(i >= candidates.length || total > target){
            return
        }
        
        total += candidates[i]
        curr.push(candidates[i])
        backtrack(i,curr,total)
        total -= candidates[i]
        curr.pop()
        backtrack(i+1,curr,total)
    }
    
    backtrack()
    return result
};